tests/Feature/PhotosAddHandlerTestAbstract.php

```diff
@@ -0,0 +1,353 @@
+			TestCase::createUploadedFile(TestCase::SAMPLE_FILE_ORIENTATION_90)
+		);
+
+		/*
+		 * Check some Exif data
+		 */
+		$response->assertJson([
+			'size_variants' => [
+				'small' => ['width' => 480, 'height' => 360],
+				'medium' => ['width' => 1440, 'height' => 1080],
+				'original' => ['width' => 2016,	'height' => 1512],
+			],
+		]);
+	}
+
+	/**
+	 * Tests auto-orientation with rotation 180° CW.
+	 *
+	 * @return void
+	 */
+	public function testUploadWithAutoOrientation180(): void
+	{
+		$this->assertHasExifToolOrSkip();
+
+		$response = $this->photos_tests->upload(
+			TestCase::createUploadedFile(TestCase::SAMPLE_FILE_ORIENTATION_180)
+		);
+
+		/*
+		 * Check some Exif data
+		 */
+		$response->assertJson([
+			'size_variants' => [
+				'small' => ['width' => 480, 'height' => 360],
+				'medium' => ['width' => 1440, 'height' => 1080],
+				'original' => ['width' => 2016,	'height' => 1512],
+			],
+		]);
+	}
+
+	/**
+	 * Tests auto-orientation with rotation 270° CW.
+	 *
+	 * @return void
+	 */
+	public function testUploadWithAutoOrientation270(): void
+	{
+		$this->assertHasExifToolOrSkip();
+
+		$response = $this->photos_tests->upload(
+			TestCase::createUploadedFile(TestCase::SAMPLE_FILE_ORIENTATION_270)
+		);
+
+		/*
+		 * Check some Exif data
+		 */
+		$response->assertJson([
+			'size_variants' => [
+				'small' => ['width' => 480, 'height' => 360],
+				'medium' => ['width' => 1440, 'height' => 1080],
+				'original' => ['width' => 2016,	'height' => 1512],
+			],
+		]);
+	}
+
+	/**
+	 * Tests auto-orientation with horizontal mirroring.
+	 *
+	 * @return void
+	 */
+	public function testUploadWithAutoOrientationHFlip(): void
+	{
+		$this->assertHasExifToolOrSkip();
+
+		$response = $this->photos_tests->upload(
+			TestCase::createUploadedFile(TestCase::SAMPLE_FILE_ORIENTATION_HFLIP)
+		);
+
+		/*
+		 * Check some Exif data
+		 */
+		$response->assertJson([
+			'size_variants' => [
+				'small' => ['width' => 480, 'height' => 360],
+				'medium' => ['width' => 1440, 'height' => 1080],
+				'original' => ['width' => 2016,	'height' => 1512],
+			],
+		]);
+	}
+
+	/**
+	 * Tests auto-orientation with vertial mirroring.
+	 *
+	 * @return void
+	 */
+	public function testUploadWithAutoOrientationVFlip(): void
+	{
+		$this->assertHasExifToolOrSkip();
+
+		$response = $this->photos_tests->upload(
+			TestCase::createUploadedFile(TestCase::SAMPLE_FILE_ORIENTATION_VFLIP)
+		);
+
+		/*
+		 * Check some Exif data
+		 */
+		$response->assertJson([
+			'size_variants' => [
+				'small' => ['width' => 480, 'height' => 360],
+				'medium' => ['width' => 1440, 'height' => 1080],
+				'original' => ['width' => 2016,	'height' => 1512],
+			],
+		]);
+	}
+
+	public function testPNGUpload(): void
+	{
+		$photo = static::convertJsonToObject($this->photos_tests->upload(
+			TestCase::createUploadedFile(TestCase::SAMPLE_FILE_PNG)
+		));
+		static::assertStringEndsWith('.png', $photo->size_variants->original->url);
```

Comment:
Correct, given the size only the smallest thumb (200px x 200px) is generated and our thumbnails are always JPGs. Hence, there is nothing to test here.

What shall I do? Upload larger test images with at least 3840px x  2160px (double the size of 1920x1080) such that all variants are generated? Of course, we should do it for all image types, i.e. PNG, TIFF, WEBP, ...

Reply:
I think there may still be value in checking just the `thumb`. As you said "our thumbnails are always JPGs" -- and that's something we _should_ be testing for, a it's not something obvious in case of, say, a PNG source.